### PR TITLE
Added SonarQube support.

### PR DESCRIPTION
**Pull request by jelaniharris**
Take a look at the SonarQube results for this project here: https://sonar.ei-app.com/dashboard?id=fssk-node

**To Run**
1. Be sure to have the environment variable SONARQUBE_LOGIN_TOKEN setup in your environment. This key is stored on PassPack. (Eventually, this will be stored in CircleCI and you won't need it locally.)
2. Run the tests to generate the coverage files (LCOV) for the two applications
```
docker exec -it fssk-node-client npm test-coverage
docker exec -it fssk-node-server npm test-coverage
```
3. Run the sonarqube.sh utility script to start the sonar-scanner process from the docker container, which will analyze the code and send the results to the SonarQube instance running at https://sonar.ei-app.com/
```
./util/sonarqube.sh
```

Once we get a CircleCI build going, I can move the additional docker service "sonarqube-scanner" into it's own docker compose yml, (probably docker-compose-test.yml specifically for CircleCI tests). This is so that developers aren't creating a sonarqube-scanner container locally that will only be used in CI.

### CODE REVIEW

**Comment by sbrudz** [comment]
Cool! We got good grades, too!  Need more test coverage, though.

The code smells don't seem right, though.  It seems like SonarQube isn't recognizing the BlueBird promises returned by the Bookshelf model methods as real promises.

I'll hopefully have a PR for the Circle CI build ready later today.

**Comment by jelaniharris** [comment]
So I did some research and figured out that this wasn't the first time that people have had the BlueBird non-native Promise issue with SonarQube (which in turn uses SonarTS).

And someone else provided a fix that was merged into SonarTS in June.
https://github.com/SonarSource/SonarTS/issues/659

I updated the plugin for SonarTS on the SonarQube server which got rid of all of those  promise related code smells, but because of the update it found other things wrong instead ... and the score isn't so great anymore 😢. 

I'll release another PR (or should I just add to this one?) with these errors fixed ... hopefully. I _might_ end up ignoring the rules it flagged because those are actually legitimate uses of the word "password" in a string.

**Comment by jelaniharris** [comment]
Well that was a fun lesson in trial and error. I really need to find a way to make this much easier.

1. I made the SonarQube sever ignore the rule (typescript:S2068) about the "possibly hard-coded password" for this project. Their solution for this rule was just setting it equal to a function's return results - which is dumb for us because we're just setting a string to explain why that password is invalid. I'd like to think we're good enough to NOT hard-code a password so ignoring this shouldn't be a problem.
2. My idea to make the SonarQube Scanner run in it's own docker container in CircleCI didn't work out exactly as I planned it to. Apparently in CircleCi you cannot assign volumes to a remote docker container (https://circleci.com/docs/2.0/building-docker-images/#mounting-folders). So I wouldn't be able to just spin up the sonar-scanner docker container (via docker exec -it sonar-scanner ....), mount the source code volumes, and tell it to do the scan. I also didn't think it would be good practice to start copying entire project folders across containers, or to use a machine. So I settled on just making the code-coverage job rely on the earthlinginteractive/sonar-scanner docker image (java 8, sonarqube, node 8), checking out the code, and restoring the workspace with the coverage files on it. Whew.
3. I ran into an issue with the sonar-scanner where it couldn't find typescript (even if it was installed globally via NPM). This doesn't seem to be an old issue, but the workaround was "just install it locally". Because the base directory for this project doesn't have any node_modules for the scanner to search for typescript, I had to install it myself in the image and somehow it still finds it?? 😕 
screen shot 2018-08-30 at 1 40 34 pm
Fun times.
4. There still seems to be a problem with the sonar-scanner being unable to change the pull request status for a pull request (this one for example) - but I think that's because the earthlingbot doesn't have the ability to change that for this project or it's not setup to depend on the SonarQube passing before validation? Not sure.

Apart from that, I think this is ready. 🤞

**Comment by sbrudz** [comment]
@jelaniharris I gave earthlingbot write permissions to this repo, so that should fix item 4 on your list.

File: .circleci/config.yml [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+          key: dependency-cache-{{ checksum "server/package.json" }}
+      - restore_cache:
+          key: dependency-cache-{{ checksum "client/package.json" }}
+      - attach_workspace:
+          at: .
+      - run:
+          name: Analyze data for SonarQube
+          command: util/sonarqube.sh
+  full_coverage:
+    <<: *code_coverage
```
Comment (sbrudz) [inline .circleci/config.yml]: What's the difference between code_coverage and full_coverage?

**Comment by jelaniharris** [comment]
*Ahem* Closing this pull request, while I make a new cleaner branch with less ... rebase mistakes.

**Pull request closed without merging**